### PR TITLE
document: replace authors by contribution field

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Adds a new pipe to transform contribution field.
* Adds licence.
* Adds a new service remoteTypeahead.
* Adds a new class for custom mef typeahead.
* Adds Record interface on result of getRecords.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>
Co-Authored-by: Johnny Mariéthoz <johnny.mariethoz@rero.ch>
Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Dependency

- https://github.com/rero/rero-ils/pull/1129

## How to test?

- Check the new field on brief and detail view (public and prof)

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: projects/admin/src/app/pipe/birth-date.pipe.ts [hunk 1]
```diff
@@ -15,13 +15,28 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ import { Pipe, PipeTransform } from '@angular/core';
 
-@Pipe({
+ @Pipe({
```
Comment (zannkukai) [inline projects/admin/src/app/pipe/birth-date.pipe.ts:36]: cosmetic : not required space before `@Pipe`

File: projects/admin/src/app/pipe/contribution-format.pipe.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   * @return string
+   */
+  transform(contribution: any, withRole = false, fallbackLanguage = 'fr'): any {
+    let agent = null;
+    if (contribution.agent) {
+      agent = this._transformAgent(contribution.agent, fallbackLanguage);
+    }
+    if (contribution.agent && 'role' in contribution && withRole) {
+      agent += ' ' + this._transformRole(contribution.role);
+    }
```
Comment (zannkukai) [inline projects/admin/src/app/pipe/contribution-format.pipe.ts:47]: This `if` statement should be included into the previous one. Soo you could remove the first test (`contribution.agent`)

File: projects/admin/src/app/pipe/birth-date.pipe.ts (hunk 1 above)
Comment (Garfield-fr) [reply projects/admin/src/app/pipe/birth-date.pipe.ts]: done

File: projects/admin/src/app/pipe/contribution-format.pipe.ts (hunk 2 above)
Comment (Garfield-fr) [reply projects/admin/src/app/pipe/contribution-format.pipe.ts]: done

File: projects/admin/src/app/pipe/contribution-format.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * @param agentRole - Agent role(s)
+   * @return string
+   */
+  private _transformRole(agentRole: any) {
+    const roles = [];
+    agentRole.map((role: string) => {
+      roles.push(this.translate.instant(role));
+    });
+    if (roles.length > 0) {
+      return '<span class="text-secondary">(' + roles.join(', ') + ')<span>';
```
Comment (jma) [inline projects/admin/src/app/pipe/contribution-format.pipe.ts:83]: you can  use backquotes here

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  /** Custom typeahead Type */
+  private _typeaheadTypes = {
+    mef: this._injector.get(MefTypeahead)
+  };
+
+  /**
+   * Constructor
+   * @param _recordService - RecordService
+   * @param _apiService - APIService
+   * @param _injector - Injector
```
Comment (jma) [inline projects/admin/src/app/service/ui-remote-typeahead.service.ts:36]: can you put some comment about the use of this service?

File: projects/admin/src/app/pipe/contribution-format.pipe.ts (hunk 3 above)
Comment (Garfield-fr) [reply projects/admin/src/app/pipe/contribution-format.pipe.ts]: done

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts (hunk 4 above)
Comment (Garfield-fr) [reply projects/admin/src/app/service/ui-remote-typeahead.service.ts]: done

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+   * @param options - remote typeahead options
+   * @param value - formControl value i.e. $ref value
+   * @returns Observable of string - html template representation of the value.
+   */
+  getValueAsHTML(options: any, value: string): Observable<string> {
+    if (options.type in this._typeaheadTypes) {
+      return this._typeaheadTypes[options.type].getValueAsHTML(options, value);
+    } else {
+      return super.getValueAsHTML(options, value);
+    }
```
Comment (zannkukai) [inline projects/admin/src/app/service/ui-remote-typeahead.service.ts:58]: COSMETIC : I love the ternary instruction. I find them much more easier to read than an if..then..else statement :
```
return (options.hasOwnProperty('type') && options.type in this._typeaheadTypes)
  ? this._typeaheadTypes[options.type].getValueAsHTML(options, value)
  : super.getValueAsHTML(options, value);
```

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+   * @param options - remote typeahead options
+   * @param query - search query to retrieve the suggestions list
+   * @param numberOfSuggestions - the max number of suggestion to return
+   * @returns - an observable of the list of suggestions.
+   */
+  getSuggestions(options: any, query: string, numberOfSuggestions: number): Observable<Array<SuggestionMetadata>> {
+    if (options.type in this._typeaheadTypes) {
+      return this._typeaheadTypes[options.type].getSuggestions(options, query, numberOfSuggestions);
+    } else {
+      super.getSuggestions(options, query, numberOfSuggestions);
```
Comment (zannkukai) [inline projects/admin/src/app/service/ui-remote-typeahead.service.ts:72]: missing return statement ? As above a ternary instruction is, for my point of view, much more clear.

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts (hunk 5 above)
Comment (Garfield-fr) [reply projects/admin/src/app/service/ui-remote-typeahead.service.ts]: done

File: projects/admin/src/app/service/ui-remote-typeahead.service.ts (hunk 6 above)
Comment (Garfield-fr) [reply projects/admin/src/app/service/ui-remote-typeahead.service.ts]: done

**Review by lauren-d — APPROVED** [review]

**Pull request merged**